hack/README.md

```diff
@@ -0,0 +1,215 @@
+
+Example:
+```sh
+for item in "${items}"; do
+    if [[ -d "${item}" ]]; then
+        some_action "${item}"
+    else
+        some_other_action "${item}"
+    fi
+done
+```
+
+### Variable Expansion
+Quote variables, prefer `"${var}"` when possible
+
+Don't brace-quote single-character shell special characters or positional parameters unless necessary or if the brace-quotes aids in readability.
+
+Example:
+```sh
+echo "Positional parameters: $1 $2 $3 ..."
+echo "Special characters: $! $- $_ $? $# $* $@ $$ ..."
+
+echo "Braces necessary for multi-digit positional parameters: ${100}"
+
+echo "Braces useful for being explicit: ${1}0 is better than $10"
+
+echo "Other variables should be quoted and braced: ${PATH}"
+```
+
+### Quoting
+Quote strings containing variables, command substitutions, spaces or shell meta characters unless expansion is explicitly unwanted. 
+
+Example:
+```sh
+# Quote command substitutions
+output="$(some_command_producing_result with flags)"
+
+# Quote variables
+echo "${var}"
+
+# Don't quote literal integers
+val=100
+
+# Single-quoting regex keeps it as-written
+grep -E '[^h]+(ello)'
+``` 
+
+## Other Guidelines
+
+### Shell Script Safety
+All executible shell scripts *must* set the following flags:
+```
+set -o errexit
+set -o nounset
+set -o pipefail
+```
+
+### Passing Arguments
+Always use `"$@"` over `"$*"`. `$@` and `$*` split on spaces, which will clobber arguments containing spaces. `"$*"` expands to one argument separated by spaces, which is often not at all what is required. Unless there is a good reason, use `"$@"`.
+
+### Command Substitution
+Use `$(command)` over `\`command\``. Nested back-ticks require escaping, while `$(command)` nests without problems.
+
+### Tests
+Use `[[` over `test` or `[`. `[[` doesn't allow for pathname expansions or word splitting and allows for regular expression matching.
+
+### Testing Strings
+Use `[[ -z "${stringvar}" ]]` or `[[ -n "${stringvar}" ]]` instead of comparing to `""`.
+
+### Eval
+`eval` should be avoided unless absolutely necessary. `eval` will expand variables and it may be very difficult to debug.
+
+### Pipes to While
+Piping into `while` can cause difficulties in debugging because of the implicit subshell that is generated for each loop. Pass input into the `while` loop explicitly.
+
+## Naming Conventions
+
+### Functions
+Functions should be named with lower-case letters, using underscores to separate words. Libraries/packages are separated with `::`. Parentheses after the function name are required, as is the `function` keyword before it. Braces begin on the same line the function is declared on.
+
+Example:
+```sh
+function os::subpackage::my_function_name {
+	...
+}
+```
+
+### Variable Names
+All variables that have global scope must be in `ALL_CAPS` using underscores to separate words. 
+
+Variables used in a script must be `all_lowercase` using underscores to separate words.
```

Comment:
If it's not local to a function, then it has global scope.  Do you mean that variables that are not used outside the file in which they are declared should have lower-case names (even though they have global scope)? Maybe I'm splitting hairs...